### PR TITLE
d3q27_PSM model

### PR DESCRIPTION
**Pull request by llaniewski**
Consolidated isothermal portion of J. McCullough's thermal LBM-DEM model with D. Wang's TRT collision operator and modified weighting function for the PSM (based on LBM-DEM coupling of Noble & Torczynski, 1998).

Compile options include:
- BGK or TRT collision (BGK is default, for TRT compile with _TRT)
- non-equilibirium-bounce-back (_NEBB) or superposition (_SUP) solid collision operators

### CODE REVIEW

File: models/particles/d3q27_PSM/Dynamics.c.Rt [hunk 1]
```diff
@@ -0,0 +1,1067 @@
...
+
+        }
+
+        if (ret > 1.0){ ret =  1.0;}
+        if (ret < 0.0){ ret = 0.0;}
+
+        return ret;
+}
+
+CudaDeviceFunction real_t CalcPeriodicSolid(){
```
Comment (llaniewski) [inline models/particles/d3q27_PSM/Dynamics.c.Rt]: This is obsolete (the lammps integration has periodicity, and ESYS will have) and inefficient (loops over all particles). I think you can get rid of it.

File: models/particles/d3q27_PSM/Dynamics.c.Rt [hunk 2]
```diff
@@ -0,0 +1,1067 @@
...
+                         omegaS = PV(paste("omegaS[",1:27-1,"]",sep=""))
+                         omegaF = PV("omegaF")
+
+                      C(usq, sum(u*u)*(1/c_squ))
+                      for (i in 1:27){
+                         C( feq[i],  (((u %*% t(U) * (1/c_squ))[i]) + (((u %*% t(U) * (1/c_squ))[i])*((u %*% t(U) * (1/c_squ))[i])+usq*(-1))*0.5 + 1) * d *tn[i,i])
+                      }
+	 ?>
+
+         for (FullParticleIterator p(X,Y,Z); p; ++p) {
```
Comment (llaniewski) [inline models/particles/d3q27_PSM/Dynamics.c.Rt]: The full particle iterator is inefficient. Please change this to `ParticleIterator` and check if your results are the same.

File: models/particles/d3q27_PSM/Dynamics.c.Rt [hunk 3]
```diff
@@ -0,0 +1,1067 @@
...
+                
+		    real_t dx = p.diff.x;
+		    real_t dy = p.diff.y;
+		    real_t dz = p.diff.z;
+
+		    if ((DNx>0) && ((DNx - abs(dx)) < (p.rad + 2))) {
```
Comment (llaniewski) [inline models/particles/d3q27_PSM/Dynamics.c.Rt]: These DNx, DNy, etc. things are related to periodicity, and should not be used anymore.

File: models/particles/d3q27_PSM/Dynamics.c.Rt [hunk 4]
```diff
@@ -0,0 +1,1067 @@
...
+		    sol = sol*(1/localOmegaF - 0.5) / ((1 - sol) + (1/localOmegaF - 0.5));
+		    //sol = sol*((1/localOmegaF - 0.5) + 2*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5) + 3*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5) + 4*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)) / ((1 - sol) + ((1/localOmegaF - 0.5) + 2*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5) + 3*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5) + 4*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)*(1/localOmegaF - 0.5)));
+
+         for (SyncParticleIterator p(X,Y,Z); p; ++p) {
+                if ((NodeType & NODE_BOUNDARY)==0){
+		    real_t dx = p.diff.x;
+		    real_t dy = p.diff.y;
+		    real_t dz = p.diff.z;
+
+		    if ((DNx>0) && ((DNx - abs(dx)) < (p.rad + 2))) {
```
Comment (llaniewski) [inline models/particles/d3q27_PSM/Dynamics.c.Rt]: periodicity again.

File: models/particles/d3q27_PSM/Dynamics.c.Rt [hunk 5]
```diff
@@ -0,0 +1,1067 @@
...
+			dy = DNy - abs(dy);
+		    }
+				
+		    if ((DNz>0) && ((DNz - abs(dz)) < (p.rad + 2))) {
+			dz = DNz - abs(dz);
+		    }
+
+		    real_t dist = sqrt(dx*dx + dy*dy + dz*dz);
+                    
+		    if ((dist - p.rad)<sqrt(3)) {
```
Comment (llaniewski) [inline models/particles/d3q27_PSM/Dynamics.c.Rt]: `sqrt(3)` is disambiguous (as `3` is an integer). You should use `sqrt(3.0)`, or better yet, use:
```c++
#define SQRT_3 1.732050807568877
...
if ((dist - p.rad)<SQRT_3) {
```

**Comment by ndivaira** [comment]
Changes now implemented, tested and pushed

**Pull request merged**